internal/services/topic.go

```diff
@@ -312,14 +314,55 @@ func (ts *TopicService) requestedAssignments(currentPartitions int, brokers []*s
 		return brokers[i].ID() < brokers[j].ID()
 	})
 
-	assignments := make(map[int32][]int32, int(partitions))
+	rackMap := make(map[string][]*sarama.Broker)
+	var rackNames []string
+	brokersWithRack := 0
+	for _, broker := range brokers {
+		if broker.Rack() != "" {
+			brokersWithRack++
+			if _, ok := rackMap[broker.Rack()]; !ok {
+				rackMap[broker.Rack()] = make([]*sarama.Broker, 0)
+				rackNames = append(rackNames, broker.Rack())
+			}
+			rackMap[broker.Rack()] = append(rackMap[broker.Rack()], broker)
+		}
+	}
+
+	if len(brokers) != brokersWithRack {
+		if brokersWithRack > 0 {
+			glog.Warningf("Not *all* brokers have rack assignments (%d/%d), topic %s will be created without rack awareness", brokersWithRack, len(brokers), ts.canaryConfig.Topic)
+		}
+	} else {
+		index := 0
+
+		for ;; {
+			again := false
+
+			for _, rackName := range rackNames {
+				brokerList := rackMap[rackName]
+				if len(brokerList) > 0 {
```

Comment:
can this really happen? If we are here, I guess all brokers have rack if we are here, or? Maybe the check is not needed?

Reply:
I don't understand, the condition controls the termination condition of the outer loop.

Reply:
Never mind. The rackMap is updated in the loop so the brokerList which becomes empty at some point.